src/transformers/trainer.py

```diff
@@ -1079,6 +1080,16 @@ def create_optimizer(self):
                         logger.debug(f"bitsandbytes: will optimize {module} in fp32")
                 logger.info(f"skipped: {skipped/2**20}M params")
 
+            if badam_kwargs is not None:
+                from badam import BlockOptimizer
+
+                self.optimizer = BlockOptimizer(
+                    base_optimizer=self.optimizer,
+                    named_parameters_list=list(opt_model.named_parameters()),
+                    block_prefix_list=None,
+                    **badam_kwargs,
+                )
```

Comment:
This is another indication of a peculiar pattern - why set `self.optimizer` and then overwrite here? Instead, the correct optimizer class and kwargs should be handled in `Trainer.get_optimizer_cls_and_kwargs(self.args, opt_model)`